src/server/routes/search.rs

```diff
@@ -48,17 +48,41 @@ pub async fn search(
                     .finish());
             }
 
-            let get_results = |page| {
-                results(
-                    &config,
-                    &cache,
-                    query,
-                    page,
-                    req.clone(),
-                    &params.safesearch,
+            // Closure to build a server_models::Cookie capturing local references
+            let build_cookie = || {
+                server_models::Cookie::build(
+                    &config.style,
+                    config
+                        .upstream_search_engines
+                        .clone()
+                        .into_iter()
+                        .filter_map(|engine_map| engine_map.1.then_some(engine_map.0))
+                        .collect(),
+                    config.safe_search,
                 )
             };
 
+            let cookie = req.cookie("appCookie");
+
+            // Get search settings using the user's cookie or from the server's config
+            let mut search_settings: server_models::Cookie<'_> = match cookie {
+                Some(ref cookie_value) => {
+                    match serde_json::from_str(cookie_value.value()) {
+                        Ok(cookie) => cookie,
+                        // If there's an issue parsing the cookie's value, default to the config
+                        Err(_) => build_cookie(),
+                    }
+                }
+                // If there's no cookie saved, use the server's config
+                None => build_cookie(),
+            };
```

Comment:
I believe you can use something like this here:
```rs
cookie
    .and_then(|c| serde_json::from_str(c.value()).ok())
    .unwrap_or_else(build_cookie);
```

Reply:
> I believe you can use something like this here:
> 
> ```rust
> cookie
>     .and_then(|c| serde_json::from_str(c.value()).ok())
>     .unwrap_or_else(build_cookie);
> ```

Well, also, I think you will not need to right an extra closure for this you will just need to add the code from the `build_cookie` closure into the `unwrap_or_else`. What do you think?

Reply:
You first need to convert the cookie to a serge something, don't you? Also unwrap_or_else here isn't taking a closure

Reply:
> You first need to convert the cookie to a serge something, don't you? Also unwrap_or_else here isn't taking a closure

Yes, you are right, but it you notice it carefully you will see that `build_cookie` in fact is a closure. :slightly_smiling_face: 

https://github.com/ddotthomas/websurfx/blob/a47e28587cd5e9e9bb93ce6520cb7ef35c436ae4/src/server/routes/search.rs#L52-L63

Also, I think the previous code had repetition, so we needed the `build_cookie` closure. As you can notice from the code snippet below.

https://github.com/ddotthomas/websurfx/blob/a47e28587cd5e9e9bb93ce6520cb7ef35c436ae4/src/server/routes/search.rs#L67-L78

 Now I think we don't need it anymore because there is only one instance `build_cookie` closure is being called. 
 
 ```rust
 cookie
    .and_then(|c| serde_json::from_str(c.value()).ok())
    .unwrap_or_else(build_cookie);
 ```
 
 So that's why I was suggesting to have change something like this:

```rust
cookie
    .and_then(|c| serde_json::from_str(c.value()).ok())
    .unwrap_or_else(
		server_models::Cookie::build(
            &config.style,
            config
                .upstream_search_engines
                .clone()
                .into_iter()
                .filter_map(|engine_map| engine_map.1.then_some(engine_map.0))
                .collect(),
            config.safe_search,
        )
    );
```

I hope you got my point :slight_smile: .

Reply:
Oh, I didn't see that. Yes, that's up to your code style.

Reply:
Also the code from the `build_cookie` seems a bit weird. Is that cloning and iteration needed for something specific?

Reply:
Yes, to solve the issue, the point of the pull request in the first place. To make sure the list of engines matches between different users and is part of the cookie.

Reply:
> I believe you can use something like this here:
> 
> ```rust
> cookie
>     .and_then(|c| serde_json::from_str(c.value()).ok())
>     .unwrap_or_else(build_cookie);
> ```

Ok, after testing this implementation, I find that it doesn't work as it complains about `data borrowed by the current function`. So after playing around for a while I found a solution and here it is, and I would recommend using this instead:

In the **server_models.rs** file: 

```rust
/// A named struct which is used to deserialize the cookies fetched from the client side.
#[allow(dead_code)]
#[derive(Deserialize)]
pub struct Cookie<'a> {
    /// It stores the theme name used in the website.
    pub theme: Cow<'a, str>,
    /// It stores the colorscheme name used for the website theme.
    pub colorscheme: Cow<'a, str>,
    /// It stores the user selected upstream search engines selected from the UI.
    pub engines: Cow<'a, Vec<Cow<'a, str>>>,
    /// It stores the user selected safe search level from the UI.
    pub safe_search_level: u8,
}

impl<'a> Cookie<'a> {
    /// ---Add Appropriate Documentation Here Following The Rust Convention---
    pub fn new(style: &'a Style, safe_search_level: u8, mut engines: Vec<Cow<'a, str>>) -> Self {
        // ---Add Appropriate Comment Here---
        engines.sort();

        Self {
            theme: Cow::Borrowed(&style.theme),
            colorscheme: Cow::Borrowed(&style.colorscheme),
            engines: Cow::Owned(engines),
            safe_search_level,
        }
    }
}
```

In the **search.rs** file:

```rust
let cookie_value_or_config_values: server_models::Cookie<'_> = cookie_value
                .and_then(|cookie| serde_json::from_str(cookie.value()).ok())
                .unwrap_or_else(|| {
                    server_models::Cookie::new(
                        &config.style,
                        config.safe_search,
                        config
                            .upstream_search_engines
                            .iter()
                            .filter_map(|(key, value)| {
                                value.then_some(std::borrow::Cow::Borrowed(key.as_str()))
                            })
                            .collect(),
                    )
                });
```

For more information on `Copy on Write (Cow)` data type. I would suggest going through this standard library docs:

https://doc.rust-lang.org/std/borrow/enum.Cow.html 

**Note:** In the above implementation, I have also removed the `.clone()` and replaced it with the `into_iter()` for the `search.rs`. Also, in the `server_models.rs` cookie struct I have now replaced the previous data types with Cow data type as it seems to work best with the serialize and deserialize traits. Additionally, they also tend to be more memory efficient.

Reply:
> Also the code from the `build_cookie` seems a bit weird. Is that cloning and iteration needed for something specific?

Actually, I think the `cloning` part could be removed as I suggested from the above code suggestion. While for the iteration part it is required because in the config the value is stored as a key value pair containing the engine with the option true or false. So when we parse the config option it gets parsed as a `hashmap` which is used to then filter out search engines which are not selected by the user. 

Also, you may ask then `why we need the whole hashmap for this? we can just parse it and have it globally available in the Config struct, right?` While the reasoning behind using the `hashmap` is that in the `src/templates/partials/engines_tab`, We use this `hashmap` to display the user the saved config or cookie setting to show that their settings are being applied. Which is essential for good user experience. I hope you got my point :slightly_smiling_face: